### PR TITLE
Allow creating slugs for emoji characters.

### PR DESCRIPTION
**Pull request by mokhan**
I would like to be able to generate archives for categories identified by an emoji character.

For example:

https://www.mokhan.ca/ls/💎/

The current version of this gem replaces the emoji character with a blank string:

```irb
irb(main):002:0> x = "💎"
=> "💎"
irb(main):003:0> Jekyll::Utils.slugify(x)
=> ""
irb(main):004:0> Jekyll::Utils.slugify(x, mode: nil)
=> ""
irb(main):005:0> Jekyll::Utils.slugify(x, mode: 'raw')
=> "💎"
```

This behaviour causes the generation of archives in the root of the `category` permalink. i.e `/:category/` instead of `/:category/💎/`.

My proposed solution adds a new configuration named `slug_mode` that allows jekyll users to change how slugs are generated. The `:raw` slug mode preserves the emoji character instead of replacing it with a blank string.

### CODE REVIEW

**Comment by ashmaroli** [comment]
I'd like this to be a ["configurable" option](https://github.com/jekyll/jekyll-archives/blob/master/docs/configuration.md#default-configuration) instead of forcing `:pretty` slugs on everyone.

```yaml
jekyll-archives:
  slug_mode: pretty # Modes supported by your Jekyll version.
                    # default: `nil`
```

**Comment by mokhan** [comment]
I updated the PR to read in the `config['slug_mode']` configuration as suggested. Is there anything else I can do?

**Comment by mokhan** [comment]
I apologise for not clearly stating my problem. I updated the description. I hope this helps.

**Comment by pathawks** [comment]
It might also be useful for slugify to raise a pretty loud warning if the input was not an empty string but the output is.

**Comment by ashmaroli** [comment]
> It might also be useful for slugify to raise a pretty loud warning

@pathawks Would that be something for Jekyll Core to do instead..?

File: test/test_jekyll_archives.rb [hunk 1]
```diff
@@ -116,7 +132,7 @@ class TestJekyllArchives < Minitest::Test
     end
 
     should "populate the {{ site.archives }} tag in Liquid" do
-      assert_equal 12, read_file("length.html").to_i
+      assert_equal 16, read_file("length.html").to_i
```
Comment (ashmaroli) [inline test/test_jekyll_archives.rb]: I wonder how the count here increased by `4`..

**Comment by pathawks** [comment]
> Would that be something for Jekyll Core to do instead..?
 
Yes. Sorry, I was assuming this _was_ a Jekyll Core issue.

File: test/test_jekyll_archives.rb (hunk 1 above)
Comment (mokhan) [reply test/test_jekyll_archives.rb]: Yes, the magic # does look. I'm happy to change to something more intuitive. Suggestions are welcome.

Here's what I understand:

Inside the `length.html` template the file emits the total # of archives.

https://github.com/jekyll/jekyll-archives/blob/272633c25832f841c07f617508d1c9d352504396/test/source/length.html#L4

Then an archive entry is created for each tag, category, year, month, day.

https://github.com/jekyll/jekyll-archives/blob/272633c25832f841c07f617508d1c9d352504396/lib/jekyll-archives.rb#L50-L76

In the `_posts` directory we had 3 templates: https://github.com/jekyll/jekyll-archives/tree/272633c25832f841c07f617508d1c9d352504396/test/source/_posts

And then I added one for this PR.

If my math is correct I think that should be 4 templates * 5 archive pages = 20?

:confused: :question: 

I'll continue to investigate.

Comment (mokhan) [reply test/test_jekyll_archives.rb]: My equation above was incorrect. The unique list of archive pages are:

Years: 2013, 2014, 2018
Months: 2013-08, 2014-03, 2014-08, 2018-11
Days: 2013-08-16, 2014-03-15, 2014-08-17, 2018-11-05
Categories: plugins, 💎
Tags: "Test tag", "tagged", "new",

16 comes from the 16 unique archives pages from the 4 different posts.

Comment (ashmaroli) [reply test/test_jekyll_archives.rb]: @mokhan Perfect! Thank you for the detailed explanation :ok_hand: :+1:

**Comment by ashmaroli** [comment]
@mokhan Sorry for the delay with this.
Are you still interested in this feature?
Upon revisiting, I realize that your assumption of the logic behind the implementation here is flawed.

The acceptable slugify modes have always [been *strings*](https://github.com/jekyll/jekyll/blob/07561b301b75010d5c4e447f68db4e4130f3bc26/lib/jekyll/utils.rb#L15):
```ruby
SLUGIFY_MODES = %w(raw default pretty ascii latin).freeze
```
And your implementation here presumably *works* in the tests because `:pretty` is not an acceptable slugify mode and therefore Jekyll just [returns the *downcased* input string](https://github.com/jekyll/jekyll/blob/07561b301b75010d5c4e447f68db4e4130f3bc26/lib/jekyll/utils.rb#L197-L203):
```ruby
def slugify(string, mode: nil, cased: false)
  mode ||= "default"
  return nil if string.nil?

  unless SLUGIFY_MODES.include?(mode)
    return cased ? string : string.downcase
  end

  ...
end
```

Therefore, I'll keep this PR open for a week more for you to revisit. (You are free to close this voluntarily, if you don't want to pursue this any more).

**Comment by mokhan** [comment]
> Are you still interested in this feature?

Yes.

This is still a problem:

```irb
irb(main):001:0> x = "💎"
irb(main):002:0> Jekyll::Utils.slugify(x)
=> ""
irb(main):003:0> Jekyll::Utils.slugify(x, mode: 'pretty')
=> ""
irb(main):004:0> Jekyll::Utils.slugify(x, mode: 'raw')
=> "💎"
irb(main):005:0> Jekyll::VERSION
=> "3.8.6"
```

If people want to generate a unique slug, this MR allows them to inject a custom slug mode. 

> Upon revisiting, I realize that your assumption of the logic behind the implementation here is flawed.
> 
> The acceptable slugify modes have always [been _strings_](https://github.com/jekyll/jekyll/blob/07561b301b75010d5c4e447f68db4e4130f3bc26/lib/jekyll/utils.rb#L15):
> 
> ```ruby
> SLUGIFY_MODES = %w(raw default pretty ascii latin).freeze
> ```
> 
> And your implementation here presumably _works_ in the tests because `:pretty` is not an acceptable slugify mode and therefore Jekyll just [returns the _downcased_ input string](https://github.com/jekyll/jekyll/blob/07561b301b75010d5c4e447f68db4e4130f3bc26/lib/jekyll/utils.rb#L197-L203):
> 

Agreed. I can remove the call to `to_sym` and improve the test.

**Comment by mokhan** [comment]
Back to you @ashmaroli. :ping_pong:

**Comment by ashmaroli** [comment]
Thank you @mokhan 
@jekyllbot: merge +minor

**Pull request merged**